### ISSUE TITLE
Tune SGDRegressor's alpha for zemir_01's all-width linear-stage escalation, then run the full fit

### ISSUE THREAD

**Issue by vltnkiz** [vltnkiz/numerai#51]
Part of #46

## Question

[Spec and, if warranted, run an SGDRegressor-based `all`-width escalation for zemir_01's linear stage](https://github.com/vltnkiz/numerai/issues/50) implemented and mechanism-validated a chunked-`partial_fit` `SGDRegressor` trainer (`train_sgd` in `zemir/models.py`), but explicitly shipped with sklearn's *default*, untuned hyperparameters (`alpha=0.0001`, `learning_rate='invscaling'`, `eta0=0.01`, `power_t=0.25`) — real tuning was scoped out to this ticket.

Its sanity checks found the untuned default is not just weaker than `train_ridge`, but genuinely unstable at `all` width specifically: on an identical 300K-row dense subsample, Ridge (`alpha=3e6`) reached coef-norm 0.00267, while untuned SGD reached coef-norm **18.08 (~6,770x larger)** — a blown-up fit, not a quality gap. The same comparison at `medium` width (780, far less collinear features) showed only a ~3.5x coefficient-norm gap. That contrast is the core finding this ticket has to act on: **regularization strength needs to scale with feature width**, not reuse one flat `alpha` — the same reason `train_ridge` needed `alpha=3e6` at `all` width in the first place (see `pick_ridge_alpha.py`'s sweep, issue #49).

Needs deciding:
- **Tuning method**: reuse `train_sgd`'s own two sanity-check harnesses from #50 (medium-width full fit; all-width dense row-subsample) as the cheap proxy to sweep over, the same way `pick_ridge_alpha.py` swept Ridge's alpha — rather than tuning directly on the expensive full `all`-width fit.
- **Alpha grid**, in the *scaled* feature space `train_sgd` already applies (`(x-2)/2`) — needs its own range, not a transferred Ridge value (`SGDRegressor`'s `alpha` is a per-sample-averaged penalty, not directly comparable to Ridge's raw-scale `alpha`).
- **Whether `eta0`/`learning_rate` also need tuning**, or whether fixing `alpha` alone resolves the blowup (the invscaling schedule already decays the effective step size over time; it's not yet established whether the instability is purely an under-regularized penalty or also an overly aggressive early step size on this wide a design matrix).
- **Pass/fail bar for finally running the full `all`-width fit**, informed by what #50 already established: same-ballpark coefficient norm and mean_corr vs Ridge on the all-width subsample proxy, not exact reproduction.

If tuning converges to a stable, sane-scale fit on the all-width subsample proxy: run the full ensemble fit at `all` width via the harness (as #50 was scoped to) and report whether it fits and completes, and what it measures — unblocking #48's actual all-vs-medium comparison. If tuning *cannot* find a stable configuration even on the cheap proxy, that itself is the signal to fall back to the hand-rolled chunked closed-form Ridge design #49 set aside (still not yet ticketed — see map #46's "Not yet specified").

**Safety note from #47/#49/#50**: don't let the OS OOM-killer or the WSL VM be the backstop when probing near the memory ceiling.

**Comment by vltnkiz** [comment]
**Progress, paused for time — resuming from here next session.**

## Grilling outcome (locked in)
- Tuning method: reuse the #50 dense-subsample proxy (300K rows, seed 0, 80/20 split, plain overall Spearman) rather than the expensive full-scale fit — extended into `~/pick_sgd_alpha.py` etc. on the WSL box, same throwaway-script convention as `pick_ridge_alpha.py`.
- Alpha grid: log-spaced `1e-4`..`100`.
- `eta0`/`learning_rate`: only tune if alpha alone doesn't stabilize things.
- Pass bar vs Ridge (`alpha=3e6`, `eval_corr=0.05221`, `coef_norm=0.00267` on the subsample): coef-norm within 10x, eval corr within 0.7x.
- Escalation ladder once a config passes: (1) resource-capped `--model linear --feature-set all --linear sgd`, (2) resource-capped `--model ensemble --feature-set all --linear sgd`, both gated on the previous step actually completing (not just "didn't crash yet"). All runs use an explicit `resource.setrlimit(RLIMIT_AS, ~40GB)` — **not** the OOM-killer/WSL-crash backstop, per #47/#49/#50's safety note (two earlier full-ensemble `all`-width attempts with tuned Ridge left empty run dirs — the VM crashed before writing anything).

## Sweep results (all on WSL, `~/pick_sgd_*.py` + their `~/sgd_*.json` outputs)
1. **Alpha-only** (`eta0` at sklearn default 0.01) — interior peak at `alpha=0.01` (eval_corr 0.0192, coef_norm 0.18) but still fails: coef_ratio 67x, corr_ratio 0.37x.
2. **Eta0 at alpha=0.01**, grid `0.001`..`1` — `eta0=0.001` best (corr_ratio 0.61x) but still failing; anything `>=0.03` diverges catastrophically (coef norms in the trillions).
3. **Refined eta0**, grid `3e-5`..`0.001` — `eta0=0.0003` clears the corr bar (0.80x) but coef_ratio still 22.87x (bar is <=10x) — coef-norm stayed roughly flat (~23-32x) across the whole eta0 range; eta0 wasn't the lever for coef-norm.
4. **Joint: re-sweep alpha at fixed eta0=0.0003**, grid `0.01`..`1.0` — **found a passing config**: `alpha=0.3, eta0=0.0003` → coef_ratio 9.67x, corr_ratio 0.73x. `alpha=0.4` also passes but right at the boundary (corr_ratio 0.70x exactly). `alpha=0.5` fails both.

**Caveat worth flagging**: the passing window is narrow (~`alpha` 0.25–0.45) and both edges of the chosen point (alpha=0.3) are close to the bars, not comfortably interior — a marginal pass on a single 300K-row subsample, not a robust one. Worth keeping in mind when judging the eventual full-scale result.

## Code changes made (uncommitted, working tree — same "leftover infra" pattern as #47/#50)
- `zemir/models.py`: `train_sgd` now takes `alpha: float = 0.0001` and `eta0: float = 0.01` params (previously hardcoded to sklearn defaults) and passes them to `SGDRegressor`.
- `scripts/fit_harness.py`: added `--sgd-alpha`/`--sgd-eta0` CLI flags (mirroring `--ridge-alpha`), wired through `_build_trainers_with_sgd`, defaulting to `train_sgd`'s own defaults when omitted.
- Not yet verified end-to-end — the `--smoke --feature-set all --linear sgd --sgd-alpha 0.3 --sgd-eta0 0.0003` sanity check was mid-run when this session had to stop (killed cleanly, no partial state left behind).

## Next steps (not yet done)
1. Re-run the smoke test above to confirm the new CLI flags actually reach `train_sgd`.
2. Ladder step 1: resource-capped `python scripts/fit_harness.py --model linear --feature-set all --linear sgd --sgd-alpha 0.3 --sgd-eta0 0.0003 --drop-columns-file ~/all_width_drop_columns.json` — chunked SGD should dodge Ridge's ~47GB peak, but this is unproven at full scale.
3. If that fits and completes: ladder step 2, the full `--model ensemble` run, same flags, same memory cap.
4. If step 2 completes: report what it measures, hand off to #48's all-vs-medium comparison.
5. If either ladder step crashes or fails to fit even with tuned SGD: that's the signal to fall back to the hand-rolled chunked closed-form Ridge design #49 set aside (spin out as a new ticket, don't build it inline).

Ticket stays open/assigned — picking this back up should start at step 1 of the ladder above, not redo the tuning.

**Comment by vltnkiz** [comment]
**Progress update — ladder step 1 done, ladder step 2 blocked on a new memory issue.**

## What happened this session
- Confirmed the `--sgd-alpha`/`--sgd-eta0` CLI wiring reaches `train_sgd` (smoke test passed).
- Found and fixed two real memory-staging bugs while running ladder step 1 (uncommitted, working tree, same "leftover infra" pattern as #47/#50):
  - `harness.py`'s `fit_validation_predictions`: `X`/`y`/`era` were `del`-ed only *inside* each trainer (`train_sgd`, `train_ridge`) — the caller's own binding to the same design matrix stayed alive through the entire validation load that follows. Now cleared in the caller too, plus an explicit `pyarrow.default_memory_pool().release_unused()` right after — pyarrow's default allocator (mimalloc) doesn't return freed pages to the OS on its own, so `del`-ing alone left ~11GB of real resident memory that only this call actually freed.
  - `models.py`'s `_ScaledSGDPredictor.predict()`: cast the *entire* input to float32 in one shot — the same class of blowup `train_sgd` exists to avoid on the training side. At `all` width this alone needed ~51GB for the ~4M-row validation set. Now chunks prediction the same way training does (reuses `train_sgd`'s own `chunk_rows`, threaded through to the returned predictor).
- **Ladder step 1 (`--model linear --feature-set all --linear sgd --sgd-alpha 0.3 --sgd-eta0 0.0003`) completed successfully** after those fixes — 4,036,034 validation rows, 647 eras, cached at `runs/harness/20260831T193849Z-linear-all-sgd-a0.3-e0.0003-drop141/`.

## What's blocking ladder step 2 (`--model ensemble`)
- Every failure this session was caught cleanly by an explicit `ulimit -v` (RLIMIT_AS) cap, exactly as the memory-boundary safety note wants — never a VM crash, always a clean Python exception.
- But `ulimit -v` turned out to be measuring the wrong thing for most of what failed: pyarrow's mimalloc-backed pool reserves large virtual arenas that don't shrink even after the corresponding Python objects are freed, so `VmSize` climbs monotonically across pipeline stages (train load → SGD training → era_boost) independent of real memory pressure. Real RSS stayed in the 11–22GB range throughout (confirmed via an instrumented diagnostic script and `ps`/`free -h`) — nowhere close to the 50GB physical ceiling.
- Raising the cap (40GB → 46GB → 56GB → 62GB) fixed each failure up through ladder step 1, but ladder step 2's `era_boost` (XGBoost) stage fails *identically* — same array shape, same ~10.5MB shortfall — at both 56GB and 62GB caps. That's inconsistent with a fixed memory need; it strongly suggests XGBoost's `QuantileDMatrix` construction (`tree_method="hist"`, `n_jobs=-1`, 32 cores detected in this WSL box) sizes its own internal batching adaptively based on available address space, so it always lands just short of whatever cap it's given — making `ulimit -v` self-defeating specifically for this stage.
- Tried switching to a real, kernel-enforced cgroup v2 memory limit instead (invisible to a process's own `getrlimit()` queries, so it shouldn't trigger that scaling): `systemd-run --user --scope -p MemoryMax=...` silently didn't enforce (`memory.max` read back as `max`) despite the `memory` controller being delegated all the way down to `app.slice`, and manually joining a cgroup (`echo $$ > cgroup.procs`) hit `Permission denied`. Didn't dig further into WSL's systemd/cgroup delegation quirks — that's its own rabbit hole, separate from this ticket's actual question.

## Next steps (not yet done)
1. Either get a working RSS-based cap (proper cgroup delegation, or whatever the `--user` scope needs) so the ensemble run's safety net doesn't fight XGBoost, or find an XGBoost-side knob that stops its `QuantileDMatrix` construction from scaling with detected address space (candidates: lower `n_jobs` from -1, or the `max_quantile_blocks`/external-memory parameters visible in the traceback).
2. Once step 2 completes (or is fixed to complete): steps 4/5 from the original plan still apply — report what it measures, hand off to #48's all-vs-medium comparison; if it still can't be made to fit, that's the signal to fall back to the hand-rolled chunked closed-form Ridge design #49 set aside.

## Code changes (uncommitted, working tree, same pattern as #47/#50's leftover infra)
- `zemir/harness.py`: clears `X, y, era` in `fit_validation_predictions` after `fit_models` returns, plus `pyarrow.default_memory_pool().release_unused()`.
- `zemir/models.py`: `_ScaledSGDPredictor` now predicts in `chunk_rows`-sized slices instead of casting its whole input to float32 at once; `train_sgd` threads its own `chunk_rows` through to the returned predictor.

Ticket stays open/assigned — picking this back up should start at ladder step 2, not redo step 1.

**Comment by vltnkiz** [comment]
**Resolved.** Alpha tuning converged, ladder step 1 ran at full scale, and ladder step 2's blocker turned out to be a *different stage* with a different ceiling — spun out rather than absorbed here.

## Answer

`SGDRegressor` at `all` width is tuned and works. `alpha=0.3, eta0=0.0003` (found on the #50 dense-subsample proxy: coef_ratio 9.67x, corr_ratio 0.73x against Ridge's `alpha=3e6` — a *marginal* pass, passing window only ~`alpha` 0.25-0.45, worth remembering when judging any eventual full-scale number).

**Ladder step 1 completed at full scale**: `--model linear --feature-set all --linear sgd --sgd-alpha 0.3 --sgd-eta0 0.0003 --drop-columns-file ~/all_width_drop_columns.json` → 4,036,034 validation rows, 647 eras, 574 train eras, 3,414 features, cached at `runs/harness/20260831T193849Z-linear-all-sgd-a0.3-e0.0003-drop141/`. **The linear stage's `all`-width memory ceiling is cleared.**

**Ladder step 2 (`--model ensemble`) is blocked by an independent ceiling in `era_boost`, not by anything this ticket owns** — see the diagnosis below. Spun out to the successor ticket; the remaining full ensemble fit goes with it, the same way #49 handed its run to #50 and #50 handed its to this one.

## The `ulimit -v` question, settled

`ulimit -v` (RLIMIT_AS) really is the wrong instrument on this box — pyarrow's mimalloc pool reserves virtual arenas that never shrink, so VmSize climbs monotonically across stages independent of real pressure. But **replacing it does not make ladder step 2 fit**, and last session's follow-on hypothesis (XGBoost sizing its internal batching against available address space, so any cap is self-defeating) is **wrong**.

Built `~/rss_guard.py` instead: polls the process tree's `VmRSS` from outside at 4Hz and SIGKILLs on breach. Measures the quantity that actually competes for the 50GiB ceiling, is invisible to the child's own `getrlimit()`, and needs no cgroup delegation (which #51 found silently non-functional under WSL's `systemd-run --user`). Smoke-tested both ways — survives under limit, kills cleanly over it. **This is the memory-boundary instrument the #47/#49/#50 safety note actually wanted; recommend it replaces `ulimit -v` for all further boundary work on this map.**

## Diagnosis of the era_boost ceiling (`~/diag_era_boost_all.py`, guard at 42 GiB)

```
[     0.0s] RSS  0.20 GiB  feature list built (3414 features)
[    18.1s] RSS 20.18 GiB  train loaded (2746268 rows)
[    18.1s] RSS 20.18 GiB  X/y sliced, train_df dropped; X.dtypes[0]=int8
[    18.1s] RSS 20.18 GiB  about to call XGBRegressor.fit(X, y)
             peak RSS 42.46 GiB -- killed by guard, fit never returned
```

RSS climbed 20.9 -> 41 GiB smoothly at ~2.5 GiB/s: an *incremental* conversion, not one large allocation.

`xgboost/data.py`'s `pandas_transform_data` converts a DataFrame **column by column** (`oth_type` -> `ser.to_numpy(dtype=np.float32)` per column), appending every column's float32 copy to one list that is held simultaneously. At 3,414 columns x 2,746,268 rows that is **37.5 GiB of float32 on top of X's own int8**, projecting to a true peak of ~57.7 GiB against a 50 GiB ceiling.

**This is structurally the identical blowup #49 diagnosed in `train_ridge` (~47GB), in a different stage.** `train_xgboost` walks into it twice: once in `XGBRegressor.fit(X, y)`, and again in `model.predict(X)`, which the era-boost loop calls on every one of its 40 iterations.

**It also explains the number last session could not explain**: one column's float32 copy is `2,746,268 x 4 = 10,985,072 bytes = 10.48 MiB` — exactly the "~10.5MB shortfall" that was identical at both the 56GB and 62GB caps. It was identical because it is the *same per-column allocation* failing, wherever memory happens to run out. Nothing was scaling with address space.

## Escalation available (handed to the successor ticket)

XGBoost has a first-class streaming path: `QuantileDMatrix` built from a `DataIter` that yields row-batches, with `DataIter(release_data=True)` — documented verbatim for this case ("Set it to True if the data transformation (converting data to np.float32 type) is memory intensive"). Each batch's float32 cast is chunk-sized and freed; what stays resident is the quantized gradient index (~1 byte/entry at `max_bin=256`, ~9.4 GiB), not a float32 matrix. Plus chunked prediction in the era-boost loop, exactly as `_ScaledSGDPredictor` already does for the linear stage.

That is the same shape of trade #50 made for Ridge->SGD, and again uses a supported library API rather than a hand-roll. Not built here — it is a separate decision with its own validation bar (the map's shared-code safety net applies: `train_xgboost` drives the live submission path, so a `medium`-width baseline reproduction gates it).

## Code changes (uncommitted, working tree, same "leftover infra" pattern as #47/#50)

Unchanged from this ticket's previous comment (`train_sgd`'s `alpha`/`eta0` params, `fit_harness.py`'s `--sgd-alpha`/`--sgd-eta0`, `harness.py`'s `X/y/era` clearing + `release_unused()`, `_ScaledSGDPredictor`'s chunked predict). Plus a one-character docstring typo fix in `harness.py` (`parquet.s` -> `parquet.`).

New throwaway scripts on the WSL box, same convention as `pick_ridge_alpha.py`: `~/rss_guard.py`, `~/diag_era_boost_all.py`, with output in `~/guard_era_boost_all.log` and `~/diag_era_boost_all.log`.
